### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Wrote a section on forms. The idea is to start off with the basics of HTML forms, and then build up how to handle submissions in JavaScript. First doing `multipart`, then `JSON` and finally a lil bit about uploading files. Hope this good; reviews would be v. welcome!

### CODE REVIEW

File: content/reference/forms.md [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+5. Now that we have the `<form>` element, we can extract all values using
+   `window.FormData()`. It gives us back a special object containing all the
+   form data that we can directly pass to the `fetch()` api. It even works in
+   all browsers!
+6. Now that we have our data, we can make a request to the server. We send it as
+   an HTTP `POST` method, and attach the `body`. Depending on the result, it
+   will now either succeed or fail.
+
+__note: People used to working with DOM events might wonder why we don't use
+`e.target` instead of `document.getElementById`. Because the `'submit'` event
```
Comment (laduke) [inline content/reference/forms.md]: code^ uses querySelector (not document.getElementById)

Comment (yoshuawuyts) [reply content/reference/forms.md]: Great catch, thanks! :sparkles:

File: content/reference/forms.md [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+This guide assumes you're working from a project generated by `create-choo-app`.
+This allows us to write inline CSS, which makes examples a little simpler to
+read. However our goal is to provide you with knowledge that translates to any
+setup - even if you don't end up using Choo.
+
+## How do forms work?
+Before we dive into how forms work in Choo, let's dive into how forms work on
+pages that don't use any JavaScript. A lot of the web was designed to work
+without JavaScript, and to submit forms, you don't need any JS at all. Knowing
+what the default behavior is of forms allows us to build on top of it, rather
```
Comment (ghost) [inline content/reference/forms.md]: `what the default behaviour of forms is`

File: content/reference/forms.md [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+than trying to rewrite functionality that's already available to us.
+
+### The Form Element
+Forms are declared using the form tag. By themselves they don't do much, but
```
Comment (ghost) [inline content/reference/forms.md]: form in backticks

File: content/reference/forms.md [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+  <input id="username" name="username" type="text">
+  <label for="password">password</label>
+  <input id="password" name="password" type="password">
+  <input type="submit" value="Login">
+</form>
+```
+
+### Validating Input
+Forms come with a wide range of validation built in. Probably the biggest
+benefit is that it works on all platforms, with little effort.  It respects user
```
Comment (ghost) [inline content/reference/forms.md]: double space after period here

File: content/reference/forms.md [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+settings such as font-size, and supports screen readers out of the box.
+
+To validate the form's input fields, there's a few attributes we can use:
+- `pattern` - validate the form's input field using a Regular Expression. For
+  example `pattern="^.{1,15}$"` makes sure strings have a length of at least 1,
+  and not more than 15.
+- `required` - make sure that the field is filled in, and valid.
+- `title` - the message to display if the `pattern` attribute is invalid. This
+  is useful for users that can't read RegExes in their error messages (e.g.
+  everyone).
```
Comment (ghost) [inline content/reference/forms.md]: if this is everyone, then the sentence itself is kind of redundant

File: content/reference/forms.md [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    required
+  >
+  <input type="submit" value="Login">
+</form>
+```
+
+## Handling Form Submissions As Multipart
+So far we've seen how to create basic HTML forms with validation. This is a
+great starting point, but often we'll want to control submissions using
+javascript.
```
Comment (ghost) [inline content/reference/forms.md]: JavaScript

File: content/reference/forms.md [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+   element. Inside it we listen for the `'submit'` event by setting the
+   `onsubmit=` attribute.
+2. We create a handler for the `'submit'` event. This will fire whenever a user
+   clicks the `type="submit"` button (or an equivalent action).
+3. Before we can handle the form's `'submit'` event, we need to disable the
+   form's default behavior.
+4. When the `onsubmit` function fires, we select the form element.
+5. Now that we have the `<form>` element, we can extract all values using
+   `window.FormData()`. It gives us back a special object containing all the
+   form data that we can directly pass to the `fetch()` api. It even works in
```
Comment (ghost) [inline content/reference/forms.md]: uppercase `api`

File: content/reference/forms.md [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+   all browsers!
+6. Now that we have our data, we can make a request to the server. We send it as
+   an HTTP `POST` method, and attach the `body`. Depending on the result, it
+   will now either succeed or fail.
+
+_note: People used to working with DOM events might wonder why we don't use
+`e.target` instead of `document.querySelector`. Because the `'submit'` event can
+be triggered not only from the `type="submit"` button, it's safer and simpler to
+select the form by id rather than try and find the right parent node.  Not
```
Comment (ghost) [inline content/reference/forms.md]: uppercase `ID`

Comment (ghost) [reply content/reference/forms.md]: also double space after the period

File: content/reference/forms.md [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+## Uploading files
+This is a hard thing to write about. Uploading files is similar to
+`multipart/form-data`, but usually requires some extra features - such as
+overriding the native form controls, showing upload progress, and validation
+strategies. There's quite a bit to cover here.
+
+Instead of covering everything, we're going to share a few useful snippets.
+Because of time constraints, we can't quite write a full section about this yet.
+But we hope this is enough to help you on your way. Pull requests to add to
+this would be amazing!
```
Comment (ghost) [inline content/reference/forms.md]: perhaps link to how to contribute to this document?

**Review by ghost — DISMISSED** [review]

File: content/reference/forms.md [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+          required
+        >
+        <input type="submit" value="Login">
+      </form>
+    </body>
+  `
+
+  function onsubmit (e) {                                              // 2.
+    e.preventDefault()                                                 // 3.
+    var form = document.querySelector('#login')                        // 4.
```
Comment (sholtomaud) [inline content/reference/forms.md]: I've been using Kory Nunn's [fastn.js](https://korynunn.github.io/fastn/) previously and trying to switch to Choo. Something that is nice in fastn that I feel like is missing here is that inputs have bindings.  Doing a document.querySelector('#login') doesn't feel right, it doesn't feel choo.

Is bindings a step too far?

**In the absence of bindings -** what about something like:

```hmtl
<form id="login" onsubmit=${onsubmit} onkeypress="${oninput}">
 <input value="${state.login.username}" >
...<><>

  function oninput (e){
      let inputValue = e.target.value + e.key;
      let name = e.srcElement.name;
      let value = {};
      value[name] = inputValue;
      emit('update:input','<formNamespace>', value);
    }
```

File: content/reference/forms.md (hunk 5 above)
Comment (yoshuawuyts) [reply content/reference/forms.md]: Hehe, yeah this was more of a joke than anything else. But it looks like it doesn't always land, so it's okay to remove.

File: content/reference/forms.md (hunk 9 above)
Comment (yoshuawuyts) [reply content/reference/forms.md]: There should be an edit button at the top when we put it on the website. Changed the phrasing a little tho; hopefully that helps! (:

File: content/reference/forms.md (hunk 10 above)
Comment (yoshuawuyts) [reply content/reference/forms.md]: This feedback made me dig in a little further. There's apparently `event.currentTarget` which gives us exactly what we need! https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget

So yeah, no need for custom syntax :grin: :sparkles:

**Comment by yoshuawuyts** [comment]
Updated with all feedback! :D

**Pull request merged**